ENH: optimize: changes to make BFGS implementation more efficient.

antonior92 opened the pull request:
I was browsing `scipy.optimize` files and realised that the BFGS implementation is not implemented in the most possible efficient way. I made three modifications to make it more efficient:
1) The current implementation uses the following formula to update the hessian inverse:
![screenshot 2017-03-12 17 19 05](https://cloud.githubusercontent.com/assets/16557411/23835620/587f26dc-0749-11e7-86ad-6ce27f06a4c0.png)
which I changed to:
![screenshot 2017-03-12 17 19 10](https://cloud.githubusercontent.com/assets/16557411/23835626/64b1cd6a-0749-11e7-974a-de80a11c95fd.png)
Both formulas produce the same results and one can be manipulated to get the other. However, while the first implements a matrix multiplication, requiring `O(n^3)` operations, the second uses only `O(n^2)` operations. Making the second a better choice for implementation
2) I am using BLAS function in order to explore the Hessian symmetry. This way the matrix operations are two times more efficient.
3) I am using the heuristic described in p.143, formula (6.20), Nocedal/Wright book to get a better initial guess for the inverse hessian `H_k`. 

In order to evaluate the effect of items (1) and (2), I have written the following [jupiter notebook](https://gist.github.com/antonior92/5c06cee871006ef4e91ef4f0d13314ba), that compare the update of `H_k` with and without the modifications (1) and (2). I have two different computers and the time results are very different depending on each I am running the tests (I think it is because of different numpy configurations). Nevertheless, the modifications (1) and (2) make the calculation of the hessian much faster for every case I have tested.

The modification (3) cause a reduction in the number of iterations for some benchmarks (``beale`` benchmark) and a increase for others (``rosenbrock`` benchmark):

![screenshot 2017-03-12 20 13 55](https://cloud.githubusercontent.com/assets/16557411/23837044/87f16ff8-0760-11e7-9f02-9adc197b547e.png)

dlax commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -147,17 +147,6 @@ def test_bfgs(self):
-                        [[0, -5.25060743e-01, 4.87748473e-01],
-                         [0, -5.24885582e-01, 4.87530347e-01]],
-                        atol=1e-14, rtol=1e-7)
-
```
Why did you drop this?

antonior92 replied:
This test demands the obtained solution to be *exactly* the same as the one that was obtained before. Because of modification (3) this does not happens. The function converges within fewer iterations for this problem and for a slight different point. The point for each the function converges  fulfil  the specified tolerance, as can be verified in the test: ``assert_allclose(self.func(params), self.func(self.solution), atol=1e-6)``.

pv commented:
As it exists so that the algorithm is not accidentally changed, rather 
than removing it should be updated.

antonior92 commented:
ok, make sense to me! I will update it...

nmayorov commented:
Very solid improvement. Formatting style could be improved though. After that and recovering the test --- +1 from me.

antonior92 commented:
I have changed the code according to the comments: I reintroduce the tests I have removed, change the doctests and, also, tried to improve the style. I appreciate any feedback in this.

Test fails seems uncorrelated with the changes I have made.

nmayorov commented:
Looks good to me. 

Squash commits into one with an informative message.

antonior92 commented:
Ok! I squash the commits into one.

nmayorov commented:
Merging tomorrow unless someone has some things to say.

pv commented:
Looks good to me.
